<a href="https://colab.research.google.com/github/atlas-github/PIDM_SentimentAnalysis/blob/master/PIDM_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Import libraries

In [0]:
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import pandas as pd

#Select target page

In [0]:
url1 = "https://forum.lowyat.net/topic/3745154/all"
response1 = requests.get(url1)
response1

<Response [200]>

#Scrape comments

In [0]:
soup1 = BeautifulSoup(response1.text, "html.parser")
result1 = soup1.findAll("div",{'class':"postcolor post_text"})
result1_list = list(result1)
list_result1 = []

for i in range(len(result1_list)):
  list_result1.append(result1[i].get_text())

list_result1

['Panda very curious, if our money is protected by PIDMthen who protected PIDM?Panda cannot brain, PIDM really that rich to protect all Malaysian wealth?   ',
 'up to 250k only ',
 'QUOTE(chokia @ Oct 15 2015, 06:11 PM)up to 250k only250k still a lot of money...  where does the money come from? ',
 'Yes.  Protected.  It is like a form of insurance.  So, they should have underwriters who would cover them. ',
 'Say only, later bankrap, sorry lah. ',
 'QUOTE(Panda @ Oct 15 2015, 06:12 PM)250k still a lot of money...  where does the money come from?donation ',
 "QUOTE(chokia @ Oct 15 2015, 06:13 PM)donationdonation only 26milliondoesn't make sense, thanks for replying anyway... ",
 'QUOTE(Panda @ Oct 15 2015, 06:10 PM)Panda very curious, if our money is protected by PIDMthen who protected PIDM?Panda cannot brain, PIDM really that rich to protect all Malaysian wealth? \xa0 What do you know about PIDM and what money the protect? ',
 'PIDM under Perkasa...anything happen u can refer to Perkas

#Scrape usernames

In [0]:
soup2 = BeautifulSoup(response1.text, "html.parser")
result2 = soup2.findAll("span", {"class": "normalname"})
result2_list = list(result2)
list_result2 = []

for i in range(len(result2_list)):
  list_result2.append(result2[i].get_text())

list_result2

['Panda',
 'chokia',
 'Panda',
 'joe_mamak',
 'Matrix',
 'chokia',
 'Panda',
 'grinders',
 'lorrydriverrocks',
 'Panda',
 'grinders',
 'Special Agent',
 'Panda',
 'grinders',
 'patienceGNR',
 'fuzzy',
 'ohman',
 'neoexcaliber',
 'grinders',
 'everest',
 'hihihehe',
 'kenshin9880',
 'patienceGNR',
 'gladfly']

#Merge usernames and comments into a table

In [0]:
dataframe = list(zip(list_result2, list_result1))
dataframe1 = pd.DataFrame(dataframe, columns = ["Person", "Comment"])
dataframe1

Person                                            Comment
0              Panda  Panda very curious, if our money is protected ...
1             chokia                                   up to 250k only 
2              Panda  QUOTE(chokia @ Oct 15 2015, 06:11 PM)up to 250...
3          joe_mamak  Yes.  Protected.  It is like a form of insuran...
4             Matrix               Say only, later bankrap, sorry lah. 
5             chokia  QUOTE(Panda @ Oct 15 2015, 06:12 PM)250k still...
6              Panda  QUOTE(chokia @ Oct 15 2015, 06:13 PM)donationd...
7           grinders  QUOTE(Panda @ Oct 15 2015, 06:10 PM)Panda very...
8   lorrydriverrocks  PIDM under Perkasa...anything happen u can ref...
9              Panda  QUOTE(grinders @ Oct 15 2015, 06:20 PM)What do...
10          grinders  QUOTE(Panda @ Oct 15 2015, 06:26 PM)PIDM offic...
11     Special Agent                                       Bank Negara 
12             Panda  QUOTE(grinders @ Oct 15 2015, 06:27 PM)Thank y...
13          grinders  QUOTE(Panda @ Oct 15 2015, 06:29 PM)All the in...
14       patienceGNR  QUOTE(grinders @ Oct 15 2015, 06:30 PM)So no n...
15             fuzzy  QUOTE(patienceGNR @ Oct 15 2015, 06:38 PM)Pret...
16             ohman                                      Use your epf 
17      neoexcaliber  I keep thinking it's just a marketing tool to ...
18          grinders  QUOTE(patienceGNR @ Oct 15 2015, 06:38 PM)Pret...
19           everest  If everyone withdraw money from bank, what wil...
20          hihihehe  QUOTE(everest @ Oct 15 2015, 07:12 PM)If every...
21       kenshin9880  QUOTE(Panda @ Oct 15 2015, 06:16 PM)donation o...
22       patienceGNR  QUOTE(grinders @ Oct 15 2015, 07:09 PM) i know...
23           gladfly  QUOTE(everest @ Oct 15 2015, 07:12 PM)If every...

#Initialise afinn sentiment analyzer

In [0]:
!pip install afinn
from afinn import Afinn
af = Afinn()

     |████████████████████████████████| 61kB 2.4MB/s 
  Created wheel for afinn: filename=afinn-0.1-cp36-none-any.whl size=53453 sha256=3926df570d6e9ac4edc7a0477d5fd8d8097f08c58f04f68484f7ed4512349683
  Stored in directory: /root/.cache/pip/wheels/b5/1c/de/428301f3333ca509dcf20ff358690eb23a1388fbcbbde008b2
Successfully built afinn


#Compute sentiment scores (polarity) and labels

In [0]:
sentiment_scores = []

for i in range(len(list_result1)):
  sentiment_scores.append(af.score(list_result1[i]))

sentiment_category = ['positive' if score > 0 
                          else 'negative' if score < 0 
                              else 'neutral' 
                                  for score in sentiment_scores]

#Sentiment statistics

In [0]:
df = pd.DataFrame([list(dataframe1['Comment']), sentiment_scores, sentiment_category]).T
df.columns = ['Comment', 'sentiment_score', 'sentiment_category']
df['sentiment_score'] = df.sentiment_score.astype('float')
df_final = pd.concat([dataframe1['Person'], df], axis = 1)
df_final

Person  ... sentiment_category
0              Panda  ...           positive
1             chokia  ...            neutral
2              Panda  ...            neutral
3          joe_mamak  ...           positive
4             Matrix  ...           negative
5             chokia  ...           positive
6              Panda  ...           positive
7           grinders  ...           positive
8   lorrydriverrocks  ...            neutral
9              Panda  ...           positive
10          grinders  ...            neutral
11     Special Agent  ...            neutral
12             Panda  ...           positive
13          grinders  ...            neutral
14       patienceGNR  ...           positive
15             fuzzy  ...           positive
16             ohman  ...            neutral
17      neoexcaliber  ...           negative
18          grinders  ...           positive
19           everest  ...            neutral
20          hihihehe  ...           negative
21       kenshin9880  ...           positive
22       patienceGNR  ...           positive
23           gladfly  ...           negative

[24 rows x 4 columns]

#Compute counts of sentiment categories

In [0]:
df_final['sentiment_category'].value_counts()

positive    12
neutral      8
negative     4
Name: sentiment_category, dtype: int64

There are more positive comments than negative comments in this thread.

#Summary statistics of sentiment scores

In [0]:
df_final['sentiment_score'].describe()

count    24.000000
mean      1.833333
std       4.517278
min     -10.000000
25%       0.000000
50%       0.500000
75%       3.250000
max      10.000000
Name: sentiment_score, dtype: float64

The average and median sentiment scores are 1.833 and 0.500 respectively.

#Download file

In [0]:
from google.colab import files
df_final.to_csv("comments_lowyat.csv")
files.download("comments_lowyat.csv")